src/main/java/com/google/sps/servlets/BlobServlet.java

```diff
@@ -18,112 +22,374 @@
+    }
+    sortImg = sortImg.toLowerCase();
+
+    Query imageQuery =
+        new Query(DataUtils.IMAGE)
+            .setAncestor(projKey)
+            .addSort("utc", sortImg.equals(DataUtils.ASCENDING_SORT)
+                                ? Query.SortDirection.ASCENDING
+                                : Query.SortDirection.DESCENDING)
+            .setFilter(combinedGetFilters(request, withMasks, DataUtils.IMAGE));
+
+    PreparedQuery storedImages = datastore.prepare(imageQuery);
+
+    ArrayList<ImageInfo> imageObjects = new ArrayList<ImageInfo>();
+
+    for (Entity imageEntity : storedImages.asIterable()) {
+      String imageUrl =
+          "/blob-host?blobkey=" + (String)imageEntity.getProperty("blobkey");
+      String imageName = (String)imageEntity.getProperty("name");
+      String imageTime = (String)imageEntity.getProperty("utc");
+      ArrayList<String> imageTags =
+          (ArrayList<String>)imageEntity.getProperty("tags");
+
+      ArrayList<MaskInfo> imageMasks = new ArrayList<MaskInfo>();
+      if (withMasks) {
+        String sortMask = request.getParameter("sort-mask");
+        // Sorted in descending chronological order by default
+        if (DataUtils.isEmptyParameter(sortMask)) {
+          sortMask = DataUtils.DESCENDING_SORT;
+        }
+        sortImg = sortImg.toLowerCase();
+
+        Query maskQuery =
+            new Query(DataUtils.MASK)
+                .setAncestor(imageEntity.getKey())
+                .addSort("utc", sortMask.equals(DataUtils.ASCENDING_SORT)
+                                    ? Query.SortDirection.ASCENDING
+                                    : Query.SortDirection.DESCENDING)
+                .setFilter(combinedGetFilters(request, withMasks, DataUtils.MASK));
+
+        PreparedQuery storedMasks = datastore.prepare(maskQuery);
+
+        for (Entity maskEntity : storedMasks.asIterable()) {
+          String maskUrl =
+              "/blob-host?blobkey=" + (String)maskEntity.getProperty("blobkey");
+          String maskName = (String)maskEntity.getProperty("name");
+          String maskTime = (String)maskEntity.getProperty("utc");
+          ArrayList<String> maskTags =
+              (ArrayList<String>)maskEntity.getProperty("tags");
+          imageMasks.add(new MaskInfo(maskUrl, maskName, maskTime, maskTags));
+        }
+      }
+
+      imageObjects.add(
+          new ImageInfo(imageUrl, imageName, imageTime, imageTags, imageMasks));
+    }
+
+    Gson gson =
+        new GsonBuilder().setPrettyPrinting().disableHtmlEscaping().create();
+    String jsonImages = gson.toJson(imageObjects);
+    response.getWriter().println(jsonImages);
+  }
+
+  /**
+   * Checks if the file uploaded to Blobstore has a valid file extension.
+   * @param     {BlobKey}   blobKey   key for the file in question
+   * @param     {boolean}   isMask    mask or image
+   * @return    {boolean}
+   */
+  private boolean isValidFile(BlobKey blobKey, boolean isMask)
```

Comment:
It's weird to me that this can never return false. Given that it either succeeds or throws an exception, maybe it would be better to call it checkFileValidity() and return void.